### PR TITLE
[BUILD-373] SHIP-0021: Local Source Upload

### PR DESCRIPTION
**Pull request by otaviof**
# Changes

## Local Source Upload

This pull-request introduces the [ability to stream a local directory](https://github.com/shipwright-io/community/blob/main/ships/0021-local-source-upload.md) content into a [Shipwright `BuildRun` generated POD](https://github.com/shipwright-io/build/pull/934). Which means, users can employ Shipwright to build container images out of their local repository clone.

The following steps represents the most common use-case scenario:

1. Create a `Build`, if it does not exists already, for instance:
```bash
shp build create nodejs-ex \
    --source-url="https://github.com/otaviof/nodejs-ex.git" \
    --output-image="image-registry.openshift-image-registry.svc:5000/otaviof/nodejs-ex:latest"
```
2. Clone the repository and make local changes:
```bash
git clone https://github.com/otaviof/nodejs-ex.git && \
    cd nodejs-ex
```
3. Use the new feature to build a container image for you:
```bash
shp build upload nodejs-ex \
    --follow \
    --output-image="image-registry.openshift-image-registry.svc:5000/otaviof/nodejs-ex:demo"
```
4. Rinse and repeat the previous step!

## Log Follower and PodWatcher

The `shp build upload` can also follow the build pod logs, using the `--follow` flag, and to have the new feature sharing the same existing components, the following changes are taking place:

- `PodWatcher` and `Follower` are now managed by Params, instantiated as a factory just like the other components employed throughout the project
- `PodWatcher` modified to use a slice of "on event" functions, so others can use the same (shared) instance
- `NewFollower` modified to have a different signature, it requires the `BuildRun` qualified name (namespace and name)

The changes here overlaps with #89.

## Testing

The changes on this pull-request have unit testing in place, and it also introduces end-to-end test scenario for the new local source upload feature. For it we simulate a complete build by streaming the local changes onto the build pod, and right after asserting a successful build takes place.

The new feature needs the changes added on the controller [PR #934](https://github.com/shipwright-io/build/pull/934), so now the end-to-end testing is using the nightly build. This should take place only until the next release.

# Submitter Checklist

- [x] Includes tests if functionality changed/was added
- [ ] Includes docs if changes are user-facing
- [x] [Set a kind label on this PR](https://prow.k8s.io/command-help#kind)  
- [x] Release notes block has been filled in, or marked NONE

# Release Notes

```release-note
Introducing "shp build upload" sub-command to stream a local repository clone data onto a newly created BuildRun, so users can employ the Shipwright Build Controller to ship container images out of arbitrary content.
```

### CODE REVIEW

**Review by gabemontero — COMMENTED** [review]
Took a very preliminary pass @otaviof and only have a high level comment for now.  And I'll preface with citing that this may already be part of your thought process, even if it is not reflected in this WIP PR yet.

And I'm guessing what I'm about to say it not much of a surprise :-)

Feels like another round of iteration on this makes sense, to get even further code reuse, in addition to the sharing of `PodWatcher` and `Tail` between the new Upload Command here and the Follower command over in https://github.com/shipwright-io/cli/blob/b42be6944302d384e4fb0b24cd06f3bbcca64a4f/pkg/shp/cmd/follower/follow.go

In particular, if we add a level of pluggability so that when the Pod hits running state, we can configure /plug in / inject the streaming of content element at https://github.com/shipwright-io/cli/blob/b5eb161f13dc2ed6b49d87e5c9b64147ca18233a/pkg/shp/cmd/build/upload.go#L237-L245 into the event handler at https://github.com/shipwright-io/cli/blob/b42be6944302d384e4fb0b24cd06f3bbcca64a4f/pkg/shp/cmd/follower/follow.go#L121-L129

we can maximize code sharing wrt the Pod event handling, and the upload code could benefit from all the recent timing fixes and error edge cases from reported bugs that are currently captured in https://github.com/shipwright-io/cli/blob/b42be6944302d384e4fb0b24cd06f3bbcca64a4f/pkg/shp/cmd/follower/follow.go#L119-L244

Feels like a maybe a new parameter that can be supplied to the `Follower` that if set results in a method callback, where the method signature matches the current `u.performDataStreaming(target)`

The existing start build and buildrun logs follow would not provide a setting for this new method callback, but Upload would.

WDYT?

**Comment by otaviof** [comment]
@gabemontero, going back on the subject of reusing the Follower, please consider my latest commit here. You will find the following:

- `PodWatcher` and `Follower` managed by `Params`, instantiated as a factory by direct usage of the Param attributes
- `PodWatcher` uses as a slice of functions, so components can use a shared instance of the `PodWatcher`
- `Follower` has a different signature, and requires the `BuildRun` name (namespace and name) during construction, then during `shp build run` the `Follower` is only instantiated during `Run` (instead of `Complete` like before)

The last item here has a direct impact on #89 discussions, [since what we've discussed](https://github.com/shipwright-io/cli/pull/89#issuecomment-1014714083) became a logical decision during the latest commit.

I wonder if we should create a spin-off pull-request with the last commit, what do you think @gabemontero and @SaschaSchwarze0? (Thanks in advance!)

**Comment by SaschaSchwarze0** [comment]
> @gabemontero, going back on the subject of reusing the Follower, please consider my latest commit here. You will find the following:
> 
>     * `PodWatcher` and `Follower` managed by `Params`, instantiated as a factory by direct usage of the Param attributes
> 
>     * `PodWatcher` uses as a slice of functions, so components can use a shared instance of the `PodWatcher`
> 
>     * `Follower` has a different signature, and requires the `BuildRun` name (namespace and name) during construction, then during `shp build run` the `Follower` is only instantiated during `Run` (instead of `Complete` like before)
> 
> 
> The last item here has a direct impact on #89 discussions, [since what we've discussed](https://github.com/shipwright-io/cli/pull/89#issuecomment-1014714083) became a logical decision during the latest commit.
> 
> I wonder if we should create a spin-off pull-request with the last commit, what do you think @gabemontero and @SaschaSchwarze0? (Thanks in advance!)

I am fine taking any reasonable solution for the problem addressed in #89. I have not been as engaged as elsewhere when it comes to the structure of our CLI, and therefore have no too strong opinion.

File: pkg/shp/cmd/build/run_test.go [hunk 1]
```diff
@@ -124,7 +124,7 @@ func TestStartBuildRunFollowLog(t *testing.T) {
 		}
 		shpclientset.PrependReactor("create", "buildruns", createReactorFunc)
 		// need this reactor to handle returning our test BuildRun with whatever updates we make based on the various
-		// test bools that result in spec.state or deletion timestamp updates
+		// test boils that result in spec.state or deletion timestamp updates
```
Comment (gabemontero) [inline pkg/shp/cmd/build/run_test.go:127]: I agree bools isn't right, but not sure boils is either

maybe just remove both bools and boils and go with `tests that result in spec.state or deletion timestamp updates `

**Review by gabemontero — COMMENTED** [review]
just a nit on one of the comments you updated @otaviof 

code changes look good per our discussion earlier today and the summary from that discussion I posted in https://github.com/shipwright-io/cli/pull/89#issuecomment-1015473951

like how these different components are converging

File: pkg/shp/cmd/build/run_test.go (hunk 1 above)
Comment (otaviof) [reply pkg/shp/cmd/build/run_test.go:127]: Oops, sorry. The I end up using the spellcheck quick fix accidentally, so I'm saving it for a future PR, please.

Comment (otaviof) [reply pkg/shp/cmd/build/run_test.go:127]: Oops, sorry! I accidentally activated the spellchecker quick fix on the word, so let me revert this change and save this information for a later interaction, please.

Comment (gabemontero) [reply pkg/shp/cmd/build/run_test.go]: +1

**Review by gabemontero — APPROVED** [review]
I'm good with the changes @otaviof but since I see some test failures both with e2e and unit that could be related from your changes, I'll refrain on lgtm'ing until you get those clean

Quick hint, with the e2e failure, it looks like a hang.  If that persists, I would not be surprised if it is in one of the follow logs e2e runs.  I've noticed that the BATS stuff is not good with dealing with hangs.  I tried to document in https://github.com/shipwright-io/cli/pull/81#issuecomment-991930023 the temporary changes I had to make there to debug, where I temporarily bypassed BATS with the e2e tests and ran the `shp` commands directly to get data.

**Comment by otaviof** [comment]
> I'm good with the changes @otaviof but since I see some test failures both with e2e and unit that could be related from your changes, I'll refrain on lgtm'ing until you get those clean
> 
> Quick hint, with the e2e failure, it looks like a hang. If that persists, I would not be surprised if it is in one of the follow logs e2e runs. I've noticed that the BATS stuff is not good with dealing with hangs. I tried to document in [#81 (comment)](https://github.com/shipwright-io/cli/pull/81#issuecomment-991930023) the temporary changes I had to make there to debug, where I temporarily bypassed BATS with the e2e tests and ran the `shp` commands directly to get data.

Thanks @gabemontero! I've fixed the end-to-end tests accordingly to the refactoring and the new local source upload. Please consider.

Also, since the controller changes are merged into this pull-request, I can make it open for review officially.

/hold cancel

**Comment by gabemontero** [comment]
update the release note section in the PR's description to something other than `TODO` (i.e. announce the new function) and I'll tag for merge @otaviof

**Comment by otaviof** [comment]
> update the release note section in the PR's description to something other than `TODO` (i.e. announce the new function) and I'll tag for merge @otaviof

Thanks for the heads up, I almost forgot about it. Please consider the updated description, it should reflect all the changes present on this PR.

**Pull request merged**